core/qtserialbluetooth.cpp

```diff
@@ -302,6 +303,33 @@ static dc_status_t qt_serial_write(void *io, const void* data, size_t size, size
 	return DC_STATUS_SUCCESS;
 }
 
+static dc_status_t qt_serial_poll(void *io, int timeout)
+{
+	qt_serial_t *device = (qt_serial_t*) io;
+
+	if (!device)
+		return DC_STATUS_INVALIDARGS;
+
+#if defined(Q_OS_WIN)
+	// FIXME FIXME FIXME!! But how ?
+	// I have no idea about windows socket programming - Linus
+	// We'll just pretend it's always readable, and hope for the best
+	// Why is the windows side not using QBluetoothSocket?
+	return DC_STATUS_SUCCESS;
```

Comment:
i do not remember the reason why we are not using QBluetoothSocket on Windows.

this is the native Windows way:
https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsapoll

```
WSAPOLLFD fdarray = { 0 };
fdarray.fd = device->socket;
fdarray.events = POLLRDNORM;
	
if (WSAPoll(&fdarray, 1, timeout) == SOCKET_ERROR) {
	qDebug() << "Error polling socket:" << WSAGetLastError();
	return DC_STATUS_TIMEOUT;
}
return DC_STATUS_SUCCESS;
```

WSAPoll needs Vista or later. only compile tested with Mingw 5.3*.

looks like Subsurface is already linking ws2_32:
https://github.com/Subsurface-divelog/subsurface/blob/master/CMakeLists.txt#L236
which includes WSAPoll.

but above this line:
https://github.com/Subsurface-divelog/subsurface/blob/df5040c6aec41fb324a9712f8da8b0b079dd09aa/core/qtserialbluetooth.cpp#L18

the following must be added.
```
#ifndef _WIN32_WINNT
#define _WIN32_WINNT 0x0600
#endif
```
maybe some of the Qt includes is already doing that.

Reply:
Frankly, I have no idea. All of the Windows code has been done by others. I'm happy to get patches / pull requests from people who either know or have the time to figure these things out.
If this waits until I find the time and opportunity to work on this with my Windows Laptop (which I haven't booted in maybe half a year) then I guess we're screwed.

Reply:
Is there somebody who can test? Either Lubomir's "implement poll()" or just the "switch over to QtBluetooth on Windows too"?

I'd actually prefer (strongly) to try to switch Windows over to QtBluetooth, because it would clean that file up enormously and we'd have the same code-base. But I obviously cannot test it..

Reply:
@janmulder I think you tested BT stuff on Windows in the path? Or am I confused?
@neolit123 has only BLE devices, I think.
See my comment above about my ability to test things.

One thing that might work is to create a pull request with the code that you want tested and then send email to the developer mailing list - we have several hundred people on that list. Someone ought to have a windows box and a BT dive computer

Reply:
> @neolit123 has only BLE devices, I think.

i only have Perdix AI, which AFAIK is BTLE only and this code path will not be executed on it?

> One thing that might work is to create a pull request with the code that you want tested and then send email to the developer mailing list - we have several hundred people on that list. Someone ought to have a windows box and a BT dive computer

we had success with that in the past.
with the new GitHub actions changes i'm not sure if possible to download the installer from a CI build:
https://github.com/Subsurface-divelog/subsurface/pull/2558/checks?check_run_id=409801119
with Travis it was possible to get the download URL from the logs.

Reply:
Here's the "enhanced" pull request with this same base, but with the addition of windows direct socket code removal.

https://github.com/Subsurface-divelog/subsurface/pull/2559

Whether it works or not, who knows...

Reply:
add test